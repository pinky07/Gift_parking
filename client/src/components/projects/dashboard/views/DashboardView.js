import React from 'react';
import Toast from 'grommet/components/Toast';

import DashboardWithCriticalError from './viewComponents/DashboardWithCriticalError';
import LoadingView from './viewComponents/LoadingView';
import DashboardWithNoCycleSnaps from './viewComponents/DashboardWithNoCycleSnaps';
import DashboardWithCycleSnaps from './viewComponents/DashboardWithCycleSnaps';

import AddCycleSnapController from '../../addCycleSnap/AddCycleSnapController';
import AddReleaseSnapController from "../../addReleaseSnap/AddReleaseSnapController";
import AddIncidentsReportController from "../../addIncidentsReport/AddIncidentsReportController";

const DashboardView = ({viewModel}) => {
  const {errorMessage} = viewModel;

  if (errorMessage)
    return (<DashboardWithCriticalError errorMessage={errorMessage}/>);
  else {
    const {name} = viewModel.project;
    if (name) {
      const elements = determineElements(viewModel);
      const {cycleSnaps} = viewModel.project;
      if (cycleSnaps.length > 0) {
        return (<DashboardWithCycleSnaps viewModel={viewModel} elements={elements}/>);
      }
      else
        return (<DashboardWithNoCycleSnaps viewModel={viewModel} elements={elements}/>);
    }
    else
      return (<LoadingView/>);
  }
}

/* Determines the UI elements that will be shown: 
  - success notification after adding a cycle snap 
  - failure notification when adding a cycle snap 
  - the modal dialog for adding a cycle snap 
*/
const determineElements = (viewModel) => {
  let elements = {
    successNotification: undefined,
    failureNotification: undefined,
    addCycleSnapLayer: undefined,
    addReleaseSnapLayer: undefined,
    addIncidentsReportLayer: undefined
  };

  const {successNotificationOnAdd} = viewModel;
  if (successNotificationOnAdd)
    elements.successNotification = <Toast status='ok'>{successNotificationOnAdd}</Toast>;

  const {failureNotificationOnAdd} = viewModel;
  if (failureNotificationOnAdd)
    elements.failureNotification = <Toast status='critical'>{failureNotificationOnAdd}</Toast>;

  const {dashboardCallbacks} = viewModel;
  const {addCycleSnap} = viewModel;
  const {addReleaseSnap} = viewModel;
  const {addIncidentsReport} = viewModel;
  
  if (addCycleSnap) {
    elements.addCycleSnapLayer = <AddCycleSnapController
      projectId={viewModel.projectId}
      projectName={viewModel.project.name}
      onClose={dashboardCallbacks.onClose}
      onSubmit={dashboardCallbacks.onAddCycleSnapSubmit}/>;
  }
  
  if (addReleaseSnap) {
    elements.addReleaseSnapLayer = <AddReleaseSnapController
      projectId={viewModel.projectId}
      projectName={viewModel.project.name}
      onClose={dashboardCallbacks.onClose}
      onSubmit={dashboardCallbacks.onAddReleaseSnapSubmit}/>;
  }

  if (addIncidentsReport) {
    elements.addIncidentsReportLayer = <AddIncidentsReportController
      projectId={viewModel.projectId}
      projectName={viewModel.project.name}
      onClose={dashboardCallbacks.onClose}
      onSubmit={dashboardCallbacks.onAddIncidentsReportSubmit}/>;
  }

  return elements;
}

export default DashboardView;