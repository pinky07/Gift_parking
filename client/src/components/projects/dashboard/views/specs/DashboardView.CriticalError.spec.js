import React from 'react';
import renderer from 'react-test-renderer';

import DashboardView from '../DashboardView';

describe('DashboardView shows critical errors', () => {

    let viewModel;
    let actual;

    it('should render a critical error', () => {
        viewModel = { errorMessage: 'There was a critical error!' }
        actual = renderer.create(<DashboardView viewModel={viewModel} />).toJSON();

        expect(actual).toMatchSnapshot();
    });
});