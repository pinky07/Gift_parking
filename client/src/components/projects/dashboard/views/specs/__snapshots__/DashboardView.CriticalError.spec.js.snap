// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardView shows critical errors should render a critical error 1`] = `
<div
  className="grommetux-box grommetux-box--direction-column grommetux-box--responsive grommetux-box--pad-none"
  id={undefined}
  onClick={undefined}
  role={undefined}
  style={Object {}}
  tabIndex={undefined}
>
  <h1
    className="grommetux-heading"
  >
    Dashboard
  </h1>
  <h3>
    <svg
      aria-label="Critical"
      className="grommetux-status-icon grommetux-status-icon-critical"
      role="img"
      version="1.1"
      viewBox="0 0 24 24"
    >
      <defs>
        <mask
          id="mask-critical-0"
        >
          <g
            className="grommetux-status-icon__detail"
          >
            <rect
              fill="#fff"
              height="24"
              width="24"
              x="0"
              y="0"
            />
            <path
              d="M8,8 L16,16"
              fill="none"
              stroke="#000"
              strokeWidth="2"
            />
            <path
              d="M8,16 L16,8"
              fill="none"
              stroke="#000"
              strokeWidth="2"
            />
          </g>
        </mask>
      </defs>
      <g
        className="grommetux-status-icon__base"
        mask="url(#mask-critical-0)"
      >
        <path
          d="M12,0 L24,12 L12,24 L0,12 Z"
          stroke="none"
        />
      </g>
    </svg>
     
    <span>
      There was a critical error!
    </span>
  </h3>
</div>
`;
